runtime/storage/cache/manager_internal.go

```diff
@@ -0,0 +1,431 @@
+func NewManager(cfg *config.Config, rt *reqtrack.RequestTracker, ts *testsupport.Manager, json jsoniter.API) *Manager {
+	return &Manager{
+		cfg:     cfg,
+		rt:      rt,
+		ts:      ts,
+		json:    json,
+		clients: make(map[string]*redis.Client),
+	}
+}
+
+func (mgr *Manager) getClient(clusterName string) *redis.Client {
+	mgr.clientMu.RLock()
+	cl := mgr.clients[clusterName]
+	mgr.clientMu.RUnlock()
+	if cl != nil {
+		return cl
+	}
+
+	// Client not found; acquire a write lock and set up the client
+	mgr.clientMu.Lock()
+	defer mgr.clientMu.Unlock()
+
+	// Did we race someone else and they set up the client first?
+	if cl := mgr.clients[clusterName]; cl != nil {
+		return cl
+	}
+
+	// Are we in a test or running in Encore Cloud? If so, use the redismock library.
+	if mgr.cfg.Static.Testing || mgr.runningInEncoreCloud() {
+		cl, err := mgr.newMiniredisClient()
+		if err != nil {
+			panic(fmt.Sprintf("cache: unable to start redis mock: %v", err))
+		}
+		mgr.clients[clusterName] = cl
+		return cl
+	}
+
+	for _, rdb := range mgr.cfg.Runtime.RedisDatabases {
+		if rdb.EncoreName == clusterName {
+			cl, err := mgr.newClient(rdb)
+			if err != nil {
+				panic(fmt.Sprintf("cache: unable to create redis client: %v", err))
+			}
+			mgr.clients[clusterName] = cl
+			return cl
+		}
+	}
+
+	panic(fmt.Sprintf("cache: unknown cluster %q", clusterName))
+}
+
+func (mgr *Manager) runningInEncoreCloud() bool {
+	if mgr.cfg != nil && mgr.cfg.Runtime.EnvCloud == "encore" {
+		return true
+	}
+	return false
+}
+
+func (mgr *Manager) newClient(rdb *config.RedisDatabase) (*redis.Client, error) {
+	srv := mgr.cfg.Runtime.RedisServers[rdb.ServerID]
+	opts := &redis.Options{
+		Network:      "tcp",
+		Addr:         srv.Host,
+		Username:     srv.User,
+		Password:     srv.Password,
+		DB:           rdb.Database,
+		MinIdleConns: orDefault(rdb.MinConnections, 1),
+		PoolSize:     orDefault(rdb.MaxConnections, runtime.GOMAXPROCS(0)*10),
+	}
+	if strings.HasPrefix(srv.Host, "/") {
+		opts.Network = "unix"
+	}
+
+	if srv.ServerCACert != "" || srv.ClientCert != "" {
+		opts.TLSConfig = &tls.Config{}
+		if srv.ServerCACert != "" {
+			caCertPool := x509.NewCertPool()
+			if !caCertPool.AppendCertsFromPEM([]byte(srv.ServerCACert)) {
+				return nil, fmt.Errorf("invalid server ca cert")
+			}
+			opts.TLSConfig.RootCAs = caCertPool
+		}
+		if srv.ClientCert != "" {
+			cert, err := tls.X509KeyPair([]byte(srv.ClientCert), []byte(srv.ClientKey))
+			if err != nil {
+				return nil, fmt.Errorf("parse client cert: %v", err)
+			}
+			opts.TLSConfig.Certificates = []tls.Certificate{cert}
+		}
+	}
+
+	return redis.NewClient(opts), nil
+}
+
+func (mgr *Manager) newMiniredisClient() (*redis.Client, error) {
+	err := mgr.initTestSrv.Do(func() error {
+		var err error
+		mgr.testSrv, err = miniredis.Run()
+
+		// Periodically clean up cache keys if running in Encore Cloud.
+		if err == nil && mgr.runningInEncoreCloud() {
+			go miniredisCleanup(mgr.testSrv, 15*time.Second, 100)
```

Comment:
If everything scales back to zero after the request, will this actually run on Encore Cloud?